src/cmd/ksh93/sh/xec.c

```diff
@@ -134,6 +134,8 @@ static bool iousepipe(Shell_t *shp)
 		return(true);
 	}
 	subpipe[2] = sh_fcntl(fd,F_dupfd_cloexec,10);
+	if(subpipe[2] >= shp->gd->lim.open_max)
+		sh_iovalidfd(shp,subpipe[2]);
```

Comment:
I don't understand this logic. Why not do it this way:

```
if (!sh_iovalidfd(shp, subpipe[2]) {
    sh_close(subpipe[1]);
    sh_close(subpipe[2]);
    return false;
}
```

The `sh_iovalidfd()` function is a nop and returns true if the fd is not -1 and less than `open_max`; otherwise it resizes the array. So there isn't any need to do that test before calling that function. Also, since it returns false if there is a problem we should check for that and bail out. Of course, I don't really know what should be done in the body of that `if` block since I haven't learned enough about the ksh internals. It may be a "can't happen" situation that calls for an `abort()` (or whatever is idiomatic for ksh in such situations).